packages/langium/src/documents/document-builder.ts

```diff
@@ -76,40 +77,54 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         const validator = this.documentValidator;
         if (this.connection || forceDiagnostics) {
             diagnostics = await validator.validateDocument(document, cancelToken);
-            await interruptAndCheck(cancelToken);
             if (this.connection) {
                 // Send the computed diagnostics to VS Code.
                 this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
             }
-            await interruptAndCheck(cancelToken);
             document.state = DocumentState.Validated;
         }
         return diagnostics;
     }
 
-    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
-        this.langiumDocuments.invalidateDocument(uri);
-        const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
+    async update(changed: URI[], deleted: URI[], cancelToken = CancellationToken.None): Promise<void> {
+        // Gather deleted documents first, as they won't be available later
+        // They are needed to correctly collect all affected documents
+        const deletedDocuments = deleted.map(e => this.langiumDocuments.getOrCreateDocument(e));
```

Comment:
AFAICT we only need the URIs of the deleted documents, not their LangiumDocument instances.